contracts/azorius/LinearERC20VotingWithHatsProposalCreation.sol

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity =0.8.19;
+
+import {IVotes} from "@openzeppelin/contracts/governance/utils/IVotes.sol";
+import {LinearERC20VotingExtensible} from "./LinearERC20VotingExtensible.sol";
+import {IHats} from "../interfaces/hats/IHats.sol";
+
+/**
+ * An [Azorius](./Azorius.md) [BaseStrategy](./BaseStrategy.md) implementation that
+ * enables linear (i.e. 1 to 1) token voting, with proposal creation restricted to
+ * users wearing whitelisted Hats.
+ */
+contract LinearERC20VotingWithHatsProposalCreation is
+    LinearERC20VotingExtensible
+{
+    event HatWhitelisted(uint256 hatId);
+    event HatRemovedFromWhitelist(uint256 hatId);
+
+    IHats public hatsContract;
+
+    /** Array to store whitelisted Hat IDs. */
+    uint256[] public whitelistedHatIds;
+
+    error InvalidHatsContract();
+    error HatAlreadyWhitelisted();
+    error HatNotWhitelisted();
+
+    /**
+     * Sets up the contract with its initial parameters.
+     *
+     * @param initializeParams encoded initialization parameters: `address _owner`,
+     * `address _governanceToken`, `address _azoriusModule`, `uint32 _votingPeriod`,
+     * `uint256 _quorumNumerator`, `uint256 _basisNumerator`, `address _hatsContract`,
+     * `uint256[] _initialWhitelistedHats`
+     */
+    function setUp(bytes memory initializeParams) public override {
+        (
+            address _owner,
+            address _governanceToken,
+            address _azoriusModule,
+            uint32 _votingPeriod,
+            uint256 _quorumNumerator,
+            uint256 _basisNumerator,
+            address _hatsContract,
+            uint256[] memory _initialWhitelistedHats
+        ) = abi.decode(
+                initializeParams,
+                (
+                    address,
+                    address,
+                    address,
+                    uint32,
+                    uint256,
+                    uint256,
+                    address,
+                    uint256[]
+                )
+            );
+
+        super.setUp(
+            abi.encode(
+                _owner,
+                _governanceToken,
+                _azoriusModule,
+                _votingPeriod,
+                0, // requiredProposerWeight is zero because we only care about the hat check
+                _quorumNumerator,
+                _basisNumerator
+            )
+        );
+
+        if (_hatsContract == address(0)) revert InvalidHatsContract();
+        hatsContract = IHats(_hatsContract);
+
+        for (uint256 i = 0; i < _initialWhitelistedHats.length; i++) {
```

Comment:
Shall we require that `_initialWhitelistedHats` has at least 1 hat? Otherwise strategy will be locked

Reply:
yes we should! Currently having trouble getting this associated test written though...

Reply:
This is fixed now